spring-cloud-stream-binders/spring-cloud-stream-binder-gemfire/src/main/java/org/springframework/cloud/stream/binder/gemfire/MessageKey.java

```diff
@@ -0,0 +1,229 @@
+
+import org.springframework.messaging.Message;
+
+/**
+ * Key class for storing {@link Message messages} in a GemFire region.
+ * This class consists of the following fields:
+ * <ul>
+ *     <li>A unique sequence id generated by the client; this id
+ *         is used to determine the order of messages</li>
+ *     <li>A producer id which is a combination of the producer PID
+ *         and the timestamp of when the producer process was started.</li>
+ *     <li>An integer used as the hash code for GemFire to select a
+ *         bucket for storing the message in a partitioned region.</li>
+ * </ul>
+ *
+ * @author Patrick Peralta
+ */
+public class MessageKey implements Externalizable, Comparable<MessageKey>,
+		PartitionResolver<MessageKey, Message<?>> {
+
+	/**
+	 * Unique sequential id for message.
+	 */
+	private long sequenceId;
+
+	/**
+	 * Encoded identifier for producer. First (leftmost) 16 bits are
+	 * reserved for the PID of the producer process, rightmost 48
+	 * bits are reserved for the timestamp of when the producer
+	 * process was started.
+	 */
+	private long producerId;
+
+	/**
+	 * Object returned for {@link #getRoutingObject}.
+	 */
+	private int routingHash;
+
+	/**
+	 * Constructor for serialization.
+	 */
+	public MessageKey() {
+	}
+
+	/**
+	 * Construct a {@code MessageKey}.
+	 *
+	 * @param sequenceId unique sequence id used for ordering messages
+	 * @param producerTimestamp time stamp of when the producer process started
+	 * @param producerPid PID of producer process
+	 */
+	public MessageKey(long sequenceId, long producerTimestamp, int producerPid) {
+		this.sequenceId = sequenceId;
+		this.producerId = producerId(producerTimestamp, producerPid);
+		this.routingHash = this.hashCode();
+	}
+
+	/**
+	 * Construct a {@code MessageKey} with a specific hash for selecting the
+	 * partitioned region bucket.
+	 *
+	 * @param sequenceId unique sequence id used for ordering messages
+	 * @param producerTimestamp time stamp of when the producer process started
+	 * @param producerPid PID of producer process
+	 * @param routingHash routing hash for bucket selection
+	 */
+	public MessageKey(long sequenceId, long producerTimestamp, int producerPid, int routingHash) {
+		this.sequenceId = sequenceId;
+		this.producerId = producerId(producerTimestamp, producerPid);
+		this.routingHash = routingHash;
+	}
+
+	/**
+	 * Return a producer id consisting of the timestamp of the process startup
+	 * time for the producer and the PID of the producer.
+	 * <p>
+	 * The first (leftmost) 16 bits are reserved for the PID of the producer
+	 * process, rightmost 48 bits are reserved for the timestamp of when the
+	 * producer process was started.
+	 *
+	 * @param producerTimestamp startup timestamp for producer
+	 * @param producerPid PID for producer
+	 * @return producer id
+	 */
+	private long producerId(long producerTimestamp, int producerPid) {
+		long longPid = producerPid;
+		longPid = longPid << 48;
```

Comment:
this line can be folded in the line above. The optimizer will happily do it for you though